### ISSUE TITLE
Java collection methods that accept Object should consider type hierarchy

### ISSUE THREAD

**Issue by wmdietl** [typetools/checker-framework#2000]
The ErrorProne check [Incompatible type as argument to Object-accepting Java collections method](http://errorprone.info/bugpattern/CollectionIncompatibleType) ensures that methods like `Collection.contains` take an argument of the right type.

It would be useful to extend this check to be aware of the type hierarchy of a particular checker.
For example, from #1999:

```java
    private final Map<String, String> map;

    public @Nullable String get(@Nullable String key) {
        return map.get(key);
    }
```

`Map.get` is declared to take a `@Nullable Object` and therefore the above code passes the Nullness Checker.
However, allowing `null` isn't useful, as the key type of `map` is `@NonNull String`.

It would be useful, for all type systems, if there were at least a warning about such incorrect usage.
